File: app/main/paths.js
'use strict';
// Resolve every path the app needs, both in dev (`electron .` from the repo)
// and when packaged (AppImage/.exe/.app running off a USB stick).
const path = require('path');
const fs = require('fs');
const { app } = require('electron');

const PLATFORM = process.platform; // 'linux' | 'win32' | 'darwin'
const ARCH = process.arch; // 'x64' | 'arm64'

// Where bundled, read-only resources live (ollama binary, python runtime, assets).
function resourcesRoot() {
  if (app.isPackaged) return process.resourcesPath;
  // dev: the build scripts populate dist/
  return path.join(__dirname, '..', '..', 'dist');
}

// Writable USB-side root that holds models/ and data/ next to the binary.
// AppImage exposes APPIMAGE (the .AppImage path); fall back to the exe dir.
function portableRoot() {
  if (!app.isPackaged) return path.join(__dirname, '..', '.run');
  if (process.env.APPIMAGE) return path.dirname(process.env.APPIMAGE);
  if (PLATFORM === 'darwin') {
    // .../plan.ai.app/Contents/MacOS/plan.ai -> dir containing the .app
    return path.resolve(path.dirname(process.execPath), '..', '..', '..');
  }
  return path.dirname(process.execPath);
}

function ensureDir(p) {
  fs.mkdirSync(p, { recursive: true });
  return p;
}

// Pick the ollama binary matching the running OS/arch from resources/ollama/.
function ollamaBinary() {
  const dir = path.join(resourcesRoot(), 'ollama');
  const name = PLATFORM === 'win32' ? 'ollama.exe' : 'ollama';
  // bundle.sh extracts the chosen flavour to resources/ollama/<bin>;
  // also tolerate a bin/ subdir (some archives nest under bin/).
  for (const cand of [path.join(dir, name), path.join(dir, 'bin', name)]) {
    if (fs.existsSync(cand)) return cand;
  }
  return path.join(dir, name); // report the expected path even if missing
}

// The python interpreter inside the relocatable venv.
function venvPython() {
  const venv = path.join(resourcesRoot(), 'runtime', 'venv');
  return PLATFORM === 'win32'
    ? path.join(venv, 'Scripts', 'python.exe')
    : path.join(venv, 'bin', 'python');
}

// Open-WebUI's installed frontend dir. The wheel force-includes the built SPA at
// open_webui/frontend, but env.py defaults FRONTEND_BUILD_DIR to BASE_DIR/build
// (wrong for an installed wheel), so we resolve + pass it explicitly.
function owFrontendDir() {
  const venv = path.join(resourcesRoot(), 'runtime', 'venv');
  const candidates = [path.join(venv, 'Lib', 'site-packages', 'open_webui', 'frontend')];
  const libdir = path.join(venv, 'lib');
  try {
    for (const d of fs.readdirSync(libdir)) {
      candidates.push(path.join(libdir, d, 'site-packages', 'open_webui', 'frontend'));
    }
  } catch { /* no lib/ (e.g. windows) */ }
  for (const c of candidates) {
    if (fs.existsSync(path.join(c, 'index.html'))) return c;
  }
  return '';
}

const paths = {
  PLATFORM,
  ARCH,
  resourcesRoot,
  portableRoot,
  ollamaBinary,
  venvPython,
  owFrontendDir,
  modelsDir: () => ensureDir(path.join(portableRoot(), 'models')),
  dataDir: () => ensureDir(path.join(portableRoot(), 'data')),
  owAssets: () => path.join(resourcesRoot(), 'ow-assets'),
  rendererIndex: () => path.join(__dirname, '..', 'renderer', 'index.html'),
  preload: () => path.join(__dirname, 'preload.js'),
};

module.exports = paths;

File: launcher/src/paths.rs
//! Resource + portable paths for the supervised services (ported from
//! app/main/paths.js). The launcher exports PLANAI_RESOURCES (the prepared
//! component tree: runtime/, ollama/, ow-assets/) and, optionally,
//! PLANAI_PORTABLE_ROOT (the writable USB-side root holding models/ + data/).

use std::path::PathBuf;

pub fn resources_root() -> PathBuf {
    std::env::var_os("PLANAI_RESOURCES")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("dist"))
}

pub fn portable_root() -> PathBuf {
    if let Some(p) = std::env::var_os("PLANAI_PORTABLE_ROOT") {
        return PathBuf::from(p);
    }
    std::env::current_exe()
        .ok()
        .and_then(|e| e.parent().map(|p| p.to_path_buf()))
        .unwrap_or_else(|| PathBuf::from("."))
}

fn ensure_dir(p: PathBuf) -> PathBuf {
    let _ = std::fs::create_dir_all(&p);
    p
}

pub fn models_dir() -> PathBuf {
    ensure_dir(portable_root().join("models"))
}
pub fn data_dir() -> PathBuf {
    ensure_dir(portable_root().join("data"))
}
pub fn ow_assets() -> PathBuf {
    resources_root().join("ow-assets")
}

/// The mounted llama.cpp component's llama-server (optional "llamacpp"
/// feature). Upstream archives differ: linux/mac pack build/bin/, win is flat.
pub fn llamacpp_server() -> PathBuf {
    let root = resources_root().join("llamacpp");
    let name = if cfg!(windows) { "llama-server.exe" } else { "llama-server" };
    for c in [root.join("build").join("bin").join(name), root.join("bin").join(name)] {
        if c.exists() {
            return c;
        }
    }
    root.join(name)
}

/// The hermes component's bundled python (optional "hermes" feature).
pub fn hermes_python() -> PathBuf {
    let root = resources_root().join("hermes").join("python");
    if cfg!(windows) {
        root.join("python.exe")
    } else {
        root.join("bin").join("python3")
    }
}

/// The ollama binary for this OS, from <resources>/ollama/ (tolerating a bin/ nest).
pub fn ollama_binary() -> PathBuf {
    let dir = resources_root().join("ollama");
    let name = if cfg!(windows) { "ollama.exe" } else { "ollama" };
    for c in [dir.join(name), dir.join("bin").join(name)] {
        if c.exists() {
            return c;
        }
    }
    dir.join(name)
}

fn runtime_root() -> PathBuf {
    resources_root().join("runtime")
}

/// The python interpreter — a host venv (runtime/venv) or a relocatable
/// python-build-standalone tree (runtime/python).
pub fn venv_python() -> PathBuf {
    let rt = runtime_root();
    let mut cands: Vec<PathBuf> = Vec::new();
    if cfg!(windows) {
        cands.push(rt.join("venv").join("Scripts").join("python.exe"));
        cands.push(rt.join("python").join("python.exe"));
    } else {
        cands.push(rt.join("venv").join("bin").join("python"));
        for base in ["python", "venv"] {
            let bin = rt.join(base).join("bin");
            cands.push(bin.join("python3"));
            if let Ok(rd) = std::fs::read_dir(&bin) {
                for e in rd.flatten() {
                    let n = e.file_name().to_string_lossy().into_owned();
                    let is_py = n == "python3"
                        || (n.starts_with("python3.") && n["python3.".len()..].chars().all(|c| c.is_ascii_digit()));
                    if is_py {
                        cands.push(e.path());
                    }
                }
            }
        }
    }
    cands.iter().find(|c| c.exists()).cloned()
        .or_else(|| cands.into_iter().next())
        .unwrap_or_else(|| rt.join("python").join("bin").join("python3"))
}

fn site_packages_roots() -> Vec<PathBuf> {
    let rt = runtime_root();
    let mut roots = Vec::new();
    for base in ["venv", "python"] {
        roots.push(rt.join(base).join("Lib").join("site-packages")); // windows
        if let Ok(rd) = std::fs::read_dir(rt.join(base).join("lib")) {
            for e in rd.flatten() {
                roots.push(e.path().join("site-packages"));
            }
        }
    }
    roots
}

/// Open-WebUI's installed frontend dir (env.py's default is wrong for a wheel).
pub fn ow_frontend_dir() -> Option<PathBuf> {
    for sp in site_packages_roots() {
        let c = sp.join("open_webui").join("frontend");
        if c.join("index.html").exists() {
            return Some(c);
        }
    }
    None
}
